File: plugins/de.ovgu.featureide.examples/featureide_examples/GPL-FH-Java/features/UndirectedWithNeighbors/GPL/Graph.java
package GPL;

import java.util.Iterator;
import java.util.LinkedList;

//dja: add for performance reasons
import java.util.HashMap;
import java.util.Map;


// *************************************************************************
   
public class Graph 
{
    public LinkedList vertices;
    public static boolean isDirected = false;
      
    //dja: add for performance reasons
    private Map verticesMap;

    public Graph( ) 
    {
        vertices = new LinkedList();

	  //dja: add for performance reasons
        verticesMap = new HashMap( );

    }
 
    // Fall back method that stops the execution of programs
    public void run( Vertex s )
    {
    }

    // Adds an edge without weights if Weighted layer is not present
    public void addEdge( Vertex start,   Neighbor theNeighbor ) 
    {
        start.addEdge( theNeighbor );
        Vertex end = theNeighbor.neighbor;
        end.addEdge( new  Neighbor( start ) );
    }

        
    public void addVertex( Vertex v ) 
    {
        vertices.add( v );

	  //dja: add for performance reasons
	  verticesMap.put( v.name, v );
    }
   
    // Finds a vertex given its name in the vertices list
    public  Vertex findsVertex( String theName )
    {
        Vertex theVertex;
        
        // if we are dealing with the root
        if ( theName == null )
            return null;

	  //dja: removed for performance reasons
//        for( VertexIter vxiter = getVertices( ); vxiter.hasNext( ); )
//        {
//            theVertex = vxiter.next( );
//            if ( theName.equals( theVertex.getName( ) ) )
//            {
//               return theVertex;
//            }
//        }
//        return null;

	  //dja: add for performance reasons
	  return ( Vertex ) verticesMap.get( theName );

    }

    public VertexIter getVertices( ) 
    {
        return new VertexIter( ) 
        {
            private Iterator iter = vertices.iterator( );
            public Vertex next( ) 
            { 
                return ( Vertex )iter.next( ); 
            }
            public boolean hasNext( ) 
            { 
                return iter.hasNext(); 
            }
        };
    }

    // Finds an Edge given both of its vertices
    public  EdgeIfc findsEdge( Vertex theSource,
                    Vertex theTarget )
       {
	  //dja: performance improvement
        //for( VertexIter vertexiter = getVertices(); vertexiter.hasNext(); )
        // {
	  //	Vertex v1 = vertexiter.next( );
	  //	for( EdgeIter edgeiter = v1.getEdges(); edgeiter.hasNext(); )
        //    {
	  //          EdgeIfc theEdge = edgeiter.next();
	  //		Vertex v2 = theEdge.getOtherVertex( v1 );
        //	      if ( ( v1.getName().equals( theSource.getName() ) &&
        //    	       v2.getName().equals( theTarget.getName() ) ) ||
        //         	     ( v1.getName().equals( theTarget.getName() ) &&
        //          	 v2.getName().equals( theSource.getName() ) ) )
        //        	return theEdge;
        //    }
        //}
		Vertex v1 = theSource;
		for( EdgeIter edgeiter = v1.getEdges(); edgeiter.hasNext(); )
            {
	            EdgeIfc theEdge = edgeiter.next();
			Vertex v2 = theEdge.getOtherVertex( v1 );
      	      if ( ( v1.getName().equals( theSource.getName() ) &&
            	       v2.getName().equals( theTarget.getName() ) ) ||
                 	     ( v1.getName().equals( theTarget.getName() ) &&
                  	 v2.getName().equals( theSource.getName() ) ) )
                	return theEdge;
            }
        return null;
    }


    public void display( ) 
    {
        System.out.println( "******************************************" );
        System.out.println( "Vertices " );
        for ( VertexIter vxiter = getVertices( ); vxiter.hasNext( ) ; )
        {
            vxiter.next( ).display( );
        }

        System.out.println( "******************************************" );
    }

    // Adds an edge without weights if Weighted layer is not present
    public EdgeIfc addEdge( Vertex start,  Vertex end )
      {
	  Neighbor e = new Neighbor( end );
        addEdge( start, e );
        return e;
    }
      
}

File: plugins/de.ovgu.featureide.examples/featureide_examples/GPL-FH-CSharp/features/UndirectedWithNeighbors/Graph.cs
//layer Undirected;

//import java.util.Iterator;
using System.Collections;
using System.Collections.Generic;

//dja: Add for performance reasons
//import java.util.HashMap;
//import java.util.Map;


// *************************************************************************
   
public class Graph 
{
    public List<Vertex> vertices;
    public static bool isDirected = false;
      
    //dja: Add for performance reasons
    private Dictionary<string, Vertex> verticesMap;

    public Graph( ) 
    {
        vertices = new List<Vertex>();

	  //dja: Add for performance reasons
        verticesMap = new Dictionary<string, Vertex>( );

    }
 
    // Fall back method that stops the execution of programs
    public void run( Vertex s )
    {
    }

    // Adds an edge without weights if Weighted layer is not present
    public void AddEdge( Vertex start,   Neighbor theNeighbor ) 
    {
        start.AddEdge( theNeighbor );
        Vertex end = theNeighbor.neighbor;
        end.AddEdge( new  Neighbor( start ) );
    }

        
    public void AddVertex( Vertex v ) 
    {
        vertices.Add( v );

	  //dja: Add for performance reasons
	  verticesMap.Add( v.name, v );
    }
   
    // Finds a vertex given its name in the vertices list
    public  Vertex findsVertex( string theName )
    {
        Vertex theVertex;
        
        // if we are dealing with the root
        if ( theName == null )
            return null;

	  //dja: Removed for performance reasons
//        for( VertexIter vxiter = GetVertices( ); vxiter.hasNext( ); )
//        {
//            theVertex = vxiter.next( );
//            if ( theName.Equals( theVertex.GetName( ) ) )
//            {
//               return theVertex;
//            }
//        }
//        return null;

	  //dja: Add for performance reasons
        return (Vertex)verticesMap[theName];

    }

    public Iterator<Vertex> GetVertices( ) {
        return new Iterator<Vertex>(vertices.GetEnumerator());
    }

    // Finds an Edge given both of its vertices
    public  EdgeIfc findsEdge( Vertex theSource,
                    Vertex theTarGet )
       {
	  //dja: performance improvement
        //for( VertexIter vertexiter = GetVertices(); vertexiter.hasNext(); )
        // {
	  //	Vertex v1 = vertexiter.next( );
	  //	for( EdgeIter edgeiter = v1.GetEdges(); edgeiter.hasNext(); )
        //    {
	  //          EdgeIfc theEdge = edgeiter.next();
	  //		Vertex v2 = theEdge.GetOtherVertex( v1 );
        //	      if ( ( v1.GetName().Equals( theSource.GetName() ) &&
        //    	       v2.GetName().Equals( theTarGet.GetName() ) ) ||
        //         	     ( v1.GetName().Equals( theTarGet.GetName() ) &&
        //          	 v2.GetName().Equals( theSource.GetName() ) ) )
        //        	return theEdge;
        //    }
        //}
		Vertex v1 = theSource;
		for( Iterator<EdgeIfc> edgeiter = v1.GetEdges(); edgeiter.hasNext(); )
            {
	            EdgeIfc theEdge = edgeiter.next();
			Vertex v2 = theEdge.GetOtherVertex( v1 );
      	      if ( ( v1.GetName().Equals( theSource.GetName() ) &&
            	       v2.GetName().Equals( theTarGet.GetName() ) ) ||
                 	     ( v1.GetName().Equals( theTarGet.GetName() ) &&
                  	 v2.GetName().Equals( theSource.GetName() ) ) )
                	return theEdge;
            }
        return null;
    }


    public void display( ) 
    {
        System.Console.Out.WriteLine( "******************************************" );
        System.Console.Out.WriteLine( "Vertices " );
        for ( Iterator<Vertex> vxiter = GetVertices( ); vxiter.hasNext( ) ; )
        {
            vxiter.next( ).display( );
        }

        System.Console.Out.WriteLine( "******************************************" );
    }

    // Adds an edge without weights if Weighted layer is not present
    public EdgeIfc AddEdge( Vertex start,  Vertex end )
      {
	  Neighbor e = new Neighbor( end );
        AddEdge( start, e );
        return e;
    }
      
}
